### PR TITLE
feat: [std/http] Handle keep-alive timeout

### PR DESCRIPTION
**Pull request by keroxp**
- `Server` closes connection when read promise is timed out after keep-alive timeout.
-  Added `readTimeout` options for `serve()` `serveTLS()` `listenAndServe()` `listenAndServeTLS` `Server`
   - `readTimeout` is keep-alive timeout from server-side. Reading headers and body can be timed out.
   - No timeout specified by default.
- Made `Server` less hacky.
- Added `letTimeout` in `util/async`: function that let promise possible to be timed out.
- BREAKING CHANGE: `readRequest`' signature changed.
- BREAKING CHANGE: `listenAndServe()` `listenAndServeTLS()` now return `Server` instance and promise of serve routine.
- Added `writeRequest` and `readResponse` in `http/io`
- Added `TimeUnits` enum in `datetime/mod`
- fix: `BufReader.readLineSlice()` should throw OS error.
- renaemd: `http/mock.ts` -> `http/testing.ts`
- renamed: `Response` -> `ServerResponse`
---
- [x] Adding more tests...

### CODE REVIEW

**Comment by keroxp** [comment]
After connection closed by peer, `tp.readLine()` throws `Deno.errors.ConnectionReset` on macOS (local machine). But it returns EOF on ubuntu CI...

**Comment by keroxp** [comment]
It seems that error bubbling from read op after TCP connection closed varies between platforms. Throwing error and returning EOF is absolutely different behavior. I hope it's because of my code...

**Comment by ry** [comment]
ConnectionReset can happen on any read. I’m not sure if that should be turned into EOF...

**Comment by keroxp** [comment]
@ry It may be `BufReader`'s specific one?

File: cli/js/tests/url_test.ts [hunk 1]
```diff
@@ -213,6 +213,19 @@ unitTest(function createBadUrl(): void {
...
+    ignore: true,
+  },
+  function pathnameShouldEncodeComponents() {
+    const url = new URL("http://🦕.com/🦖?q=🎉");
+    assertEquals(url.host, "xn--lt9h.com");
+    assertEquals(url.pathname, "/%F0%9F%A6%96");
+    assertEquals(url.search, "?q=%F0%9F%8E%89");
+  }
+);
+
```
Comment (ry) [inline cli/js/tests/url_test.ts]: This seems unrelated to this PR. Tho I agree we need to get this working. Does this require ICU?

Comment (keroxp) [reply cli/js/tests/url_test.ts:228]: Sorry, what do you mean "ICU"?

Comment (keroxp) [reply cli/js/tests/url_test.ts:228]: I tired to add tests for `writeRequest` and expected url to be encoded. `url.pathname` must return URL encoded component of paths. May I separate it into another branch?

**Comment by keroxp** [comment]
On merging master, I had to revert almost all changes of #4551. Those changes inevitably conflicts this PR. I think `readRequest` shouldn't write response implicitly. That should be done by `Server`. cc @manyuanrong

**Comment by ry** [comment]
@keroxp I agree - I should have reviewed more carefully. I can revert those changes.

cc @nayeemrmn

**Comment by nayeemrmn** [comment]
> I think `readRequest` shouldn't write response implicitly

Yeah... I thought "`readRequest()` doesn't return the request object to the server on failure so it has to be handled in `readRequest()`", but we only need access to the writer which can easily be done in the server. Go ahead if you need to revert it, I'll try again later.

**Comment by keroxp** [comment]
@nayeemrmn Thank you!

**Comment by nayeemrmn** [comment]
I opened https://github.com/keroxp/deno/pull/1, maybe no need to revert.

**Comment by keroxp** [comment]
@ry Could you resume reviewing?

File: std/http/io.ts [hunk 2]
```diff
@@ -340,43 +383,151 @@ export function parseHTTPVersion(vers: string): [number, number] {
   throw new Error(`malformed HTTP version ${vers}`);
 }
 
+/** Read HTTP/1.1 request line and */
 export async function readRequest(
   conn: Deno.Conn,
-  bufr: BufReader
+  opts?: {
+    r?: BufReader;
+    w?: BufWriter;
```
Comment (ry) [inline std/http/io.ts:391]: I thought we wanted readRequest to not take a writer?

File: std/http/client.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/** Http request built on client side */
+export interface ClientRequest {
```
Comment (ry) [inline std/http/client.ts:2]: Is this only used for testing? Why not use fetch?

Comment (keroxp) [reply std/http/client.ts:2]: `fetch` support keep-alive connection?  I needed `Conn` based req/res for checking connection closed after keep-alive timeout passed. And there are a lot of raw-level http protocol writing in http tests. I want to cleanup them gradually.

File: std/http/io.ts (hunk 2 above)
Comment (keroxp) [reply std/http/io.ts:391]: Yes, it is needed to make request.w private and immutable. I agree that `readRequest` shouldn't take writer. But it's is impossible if we keep using class for ServerRequest.  Writer is needed only for responding. Here is one solution to split `ServerRequest` into request params and responder. Composing them should be done in `Server` after `readRequest.`

```ts
type Part1 = {
  url: string
  ...
}
type Part2 = {
  respond(res: ServerResponse)
}

type ServerRequest = Part1 & Part1;

function readRequest(r: Reader): Promise<Part1>

```

I think there are no reason to use class for `ServerRequest`.

**Comment by keroxp** [comment]
@ry Would you include these features in v1? I think keep-alive/read timeout is essential. But if it's difficult, I'm going to close this PR. It's hard to catch up master every week without reviews😓

**Comment by bartlomieju** [comment]
@keroxp sorry for slow review; we definitely want to land it, but it will have to wait after 1.0 is released - we'll then focus on `std/` to make it stable as well.

**Comment by keroxp** [comment]
@bardiarastin Sure, I'm still keep trying to catch up master!

**Comment by keroxp** [comment]
@bartlomieju  Sure, I'm still keep trying to catch up master!

**Comment by bartlomieju** [comment]
@keroxp sorry for asking again, but can you merge with `master`? I'll try to get this PR in into `std/0.52.0`

File: std/examples/chat/server.ts [hunk 4]
```diff
@@ -30,7 +30,8 @@ async function wsHandler(ws: WebSocket): Promise<void> {
   }
 }
 
-listenAndServe({ port: 8080 }, async (req) => {
+const port = parseInt(Deno.args[0] ?? "4601");
+listenAndServe({ port }, async (req) => {
```
[the remark this answers is not in the document]
Comment (keroxp) [reply std/examples/chat/server.ts:34]: @bartlomieju Example tests are flaky and this patch might fixes the issue. It is not related to this PR, but sometimes bothered CI.

**Comment by keroxp** [comment]
> splitting this PR into 2 separate PRs:

@bartlomieju Yeah I agree to abandon this PR. I'll try to submit new PRs later.
